src/charm.py

```diff
@@ -117,28 +119,38 @@ def _on_config_changed(self, event: ops.ConfigChangedEvent) -> None:
                 return
             self._stored.binary_url = self.config.get('binary-url')
             self._stored.docker_tag = self.config.get('docker-tag')
+            restart_needed = True
 
         # Update of polkadot service arguments requested
         if self._stored.service_args != self.config.get('service-args'):
             self.unit.status = ops.MaintenanceStatus("Updating service args")
-            utils.update_service_args(service_args_obj.service_args_string)
             self._stored.service_args = self.config.get('service-args')
+            restart_needed = True
 
         if self._stored.chain_spec_url != self.config.get('chain-spec-url'):
             self.unit.status = ops.MaintenanceStatus("Updating chain spec")
-            utils.update_service_args(service_args_obj.service_args_string)
             self._stored.chain_spec_url = self.config.get('chain-spec-url')
+            restart_needed = True
 
         if self._stored.local_relaychain_spec_url != self.config.get('local-relaychain-spec-url'):
             self.unit.status = ops.MaintenanceStatus("Updating relaychain spec")
-            utils.update_service_args(service_args_obj.service_args_string)
             self._stored.local_relaychain_spec_url = self.config.get('local-relaychain-spec-url')
+            restart_needed = True
 
         if self._stored.wasm_runtime_url != self.config.get('wasm-runtime-url'):
             self.unit.status = ops.MaintenanceStatus("Updating wasm runtime")
             utils.download_wasm_runtime(self.config.get('wasm-runtime-url'))
-            utils.update_service_args(service_args_obj.service_args_string)
             self._stored.wasm_runtime_url = self.config.get('wasm-runtime-url')
+            restart_needed = True
+
+        if restart_needed:
+            if utils.service_started():
+                utils.update_service_args(service_args_obj.service_args_string)
+                utils.restart_service()
+            else:
+                self.unit.status = ops.BlockedStatus("Service failed to start after configuration change")
+                event.defer()
+                return
```

Comment:
I would probably rename the variable to `update_service_args` or similar, since that's what it signifies more than if a restart is needed or not. If the services should restart or not will then be determined on if they are running or not. 

Also #1, shouldn't the service args be updated if the service isn't running? Seems like that might cause errors and confusion.

Also #2. why the blocked status + defer? If the service isn't running, that shouldn't affect the config change like this.

```suggestion
        if update_service_args:
            utils.update_service_args(service_args_obj.service_args_string)
            if utils.service_started():
                utils.restart_service()
```

PS. This requires removing the restart command from `utils.update_service_args()`.

Reply:
Thanks for the comments, as you suggested. I have done the following:
- Rename the variable to update_service_args
- Correct, service args should be updated when the service isn't running.
- There's no more `defer` and `BlockStatus` has been changed to `MaintenanceStatus`.